node/src/config.rs

```diff
@@ -0,0 +1,343 @@
+        Ok(Config {
+            stores,
+            deployment,
+            ingestor,
+        })
+    }
+
+    /// Genrate a JSON representation of the config.
+    pub fn to_json(&self) -> Result<String> {
+        // It would be nice to produce a TOML representation, but that runs
+        // into this error: https://github.com/alexcrichton/toml-rs/issues/142
+        // and fixing it as described in the issue didn't fix it. Since serializing
+        // this data isn't crucial and only needed for debugging, we'll
+        // just stick with JSON
+        Ok(serde_json::to_string_pretty(&self)?)
+    }
+
+    pub fn primary_store(&self) -> &Shard {
+        self.stores
+            .get(PRIMARY)
+            .expect("a validated config has a primary store")
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Shard {
+    pub connection: String,
+    #[serde(default = "one")]
+    pub weight: usize,
+    #[serde(default = "zero")]
+    pub pool_size: u32,
+    pub replicas: HashMap<String, Replica>,
+}
+
+fn check_pool_size(pool_size: u32, connection: &str) -> Result<()> {
+    if pool_size < 2 {
+        Err(anyhow!(
+            "connection pool size must be at least 2, but is {} for {}",
+            pool_size,
+            connection
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+impl Shard {
+    fn validate(&mut self, opt: &Opt) -> Result<()> {
+        self.connection = shellexpand::env(&self.connection)?.into_owned();
+        if self.pool_size == 0 {
+            self.pool_size = opt.store_connection_pool_size;
+        }
+
+        check_pool_size(self.pool_size, &self.connection)?;
+        for (name, replica) in self.replicas.iter_mut() {
+            validate_name(name)?;
+            replica.validate(opt)?;
+        }
+        Ok(())
+    }
+
+    fn from_opt(opt: &Opt) -> Result<Self> {
+        let postgres_url = opt
+            .postgres_url
+            .as_ref()
+            .expect("validation checked that postgres_url is set");
+        check_pool_size(opt.store_connection_pool_size, &postgres_url)?;
+        let mut replicas = HashMap::new();
+        for (i, host) in opt.postgres_secondary_hosts.iter().enumerate() {
+            let replica = Replica {
+                connection: replace_host(&postgres_url, &host),
+                weight: opt.postgres_host_weights.get(i + 1).cloned().unwrap_or(1),
+                pool_size: opt.store_connection_pool_size,
+            };
+            replicas.insert(format!("replica{}", i + 1), replica);
+        }
+        Ok(Self {
+            connection: postgres_url.clone(),
+            weight: opt.postgres_host_weights.get(0).cloned().unwrap_or(1),
+            pool_size: opt.store_connection_pool_size,
+            replicas,
+        })
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Replica {
+    pub connection: String,
+    #[serde(default = "one")]
+    pub weight: usize,
+    #[serde(default = "zero")]
+    pub pool_size: u32,
```

Comment:
Since `0` is the default integer, you could have this just be `#[serde(default)]`, if you find it equally readable.

Reply:
Nice, didn't realize that.